File: competitive_intelligence_reports/duchenne_muscular_dystrophy/2026-05-12_1235/sources/capture_sources.py
from pathlib import Path
from playwright.sync_api import sync_playwright


RUN = Path("/Users/justinyu/Desktop/linkedin-posts/competitive_intelligence_reports/duchenne_muscular_dystrophy/2026-05-12_1235")
DATE_DIR = RUN / "screenshots" / "date-verification"
EVIDENCE_DIR = RUN / "screenshots" / "evidence"

SOURCES = [
    {
        "id": "01",
        "name": "Entrada Therapeutics",
        "url": "https://www.globenewswire.com/news-release/2026/05/07/3289797/0/en/Entrada-Therapeutics-Announces-Positive-Topline-Results-from-Cohort-1-of-Participants-with-Duchenne-Muscular-Dystrophy-Treated-with-ENTR-601-44-in-Phase-1-2-ELEVATE-44-201-Study.html",
        "date": "May 07, 2026",
        "evidence": [
            "Cohort 1 demonstrated an increase of 2.36% in dystrophin",
            "Statistically significant and potentially differentiated improvement in treated participants versus placebo in Time to Rise velocity",
        ],
    },
    {
        "id": "02",
        "name": "Solid Biosciences",
        "url": "https://www.globenewswire.com/news-release/2026/05/07/3289921/0/en/Solid-Biosciences-Doses-First-Participant-in-Phase-3-IMPACT-DUCHENNE-Clinical-Trial-Evaluating-SGT-003-in-Duchenne-Muscular-Dystrophy.html",
        "date": "May 07, 2026",
        "evidence": [
            "first participant has been dosed in IMPACT DUCHENNE",
            "46 participants have been dosed with SGT-003",
        ],
    },
    {
        "id": "03",
        "name": "Sarepta Therapeutics",
        "url": "https://www.sec.gov/Archives/edgar/data/873303/000119312526208966/srpt-ex99_1.htm",
        "date": "May 6, 2026",
        "evidence": [
            "Net product revenues for the first quarter 2026 totaled $330.5 million",
            "Completed submission of sNDA for AMONDYS 45 and VYONDYS 53",
        ],
    },
    {
        "id": "04",
        "name": "Capricor Therapeutics",
        "url": "https://www.capricor.com/investors/news-events/press-releases/detail/343/capricor-therapeutics-takes-legal-action-to-protect-patient",
        "date": "May 07, 2026",
        "evidence": [
            "filed a lawsuit against Nippon Shinyaku",
            "Priority Review, with a target PDUFA action date of August 22, 2026",
        ],
    },
    {
        "id": "05",
        "name": "Dyne Therapeutics",
        "url": "https://www.sec.gov/Archives/edgar/data/1818794/000119312526215720/dyn-ex99_1.htm",
        "date": "May 11, 2026",
        "evidence": [
            "Positive pre-BLA meeting completed with FDA for z-rostudirsen",
            "on track for BLA submission in Q2 2026 and potential launch in Q1 2027",
        ],
    },
    {
        "id": "06",
        "name": "Atossa Therapeutics",
        "url": "https://investors.atossatherapeutics.com/2026-05-08-Atossa-Therapeutics-Reports-First-Quarter-2026-Financial-Results-and-Provides-a-Corporate-Update",
        "date": "May 8, 2026",
        "evidence": [
            "improved muscle strength, increased lean mass",
            "Orphan Drug Designation to (Z)-endoxifen for the treatment of DMD",
        ],
    },
    {
        "id": "07",
        "name": "Tenaya Therapeutics",
        "url": "https://investors.tenayatherapeutics.com/news-releases/news-release-details/tenaya-therapeutics-reports-first-quarter-2026-financial-results",
        "date": "May 06, 2026",
        "evidence": [
            "TN-301 treatment at doses as low as 3 mg/kg improved grip strength",
            "TN-301 was granted both Rare Pediatric Disease Designation and Orphan Drug Designation",
        ],
    },
]


HIGHLIGHT_JS = r"""
(phrases) => {
  const style = document.createElement('style');
  style.textContent = `
    mark.ci-highlight {
      background: #d7ff5f !important;
      color: #10120f !important;
      outline: 4px solid #ff3b30 !important;
      outline-offset: 3px !important;
      border-radius: 2px !important;
      box-shadow: 0 0 0 3px rgba(255,255,255,.85) !important;
      padding: 0 2px !important;
    }
    .ci-box-highlight {
      outline: 5px solid #ff3b30 !important;
      outline-offset: 5px !important;
      background: rgba(215,255,95,.32) !important;
      box-shadow: 0 0 0 4px rgba(255,255,255,.9) !important;
    }
    .ci-hide, [id*="cookie" i], [class*="cookie" i], [id*="onetrust" i], [class*="onetrust" i],
    [class*="newsletter" i], [id*="newsletter" i], [class*="modal" i], [role="dialog"] {
      display: none !important;
      visibility: hidden !important;
    }
  `;
  document.head.appendChild(style);

  const skipped = new Set(['SCRIPT', 'STYLE', 'NOSCRIPT', 'MARK', 'SVG']);
  const escapeRegExp = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const normalize = (s) => s.replace(/\s+/g, ' ').trim();
  const results = [];

  function markTextNode(textNode, phrase) {
    const text = textNode.nodeValue;
    const index = text.indexOf(phrase);
    if (index === -1) return null;
    const before = document.createTextNode(text.slice(0, index));
    const mark = document.createElement('mark');
    mark.className = 'ci-highlight';
    mark.textContent = text.slice(index, index + phrase.length);
    const after = document.createTextNode(text.slice(index + phrase.length));
    textNode.parentNode.insertBefore(before, textNode);
    textNode.parentNode.insertBefore(mark, textNode);
    textNode.parentNode.insertBefore(after, textNode);
    textNode.parentNode.removeChild(textNode);
    return mark;
  }

  function markExactPhrase(phrase) {
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT, {
      acceptNode(node) {
        if (!node.nodeValue || !normalize(node.nodeValue)) return NodeFilter.FILTER_REJECT;
        if (node.parentElement && skipped.has(node.parentElement.tagName)) return NodeFilter.FILTER_REJECT;
        return NodeFilter.FILTER_ACCEPT;
      }
    });
    const textNodes = [];
    while (walker.nextNode()) textNodes.push(walker.currentNode);
    for (const node of textNodes) {
      const mark = markTextNode(node, phrase);
      if (mark) return mark;
    }
    return null;
  }

  function markContainingElement(phrase) {
    const elements = Array.from(document.body.querySelectorAll('p, li, h1, h2, h3, h4, td, div, span, font, time'));
    const matches = elements
      .filter((el) => normalize(el.innerText || '').includes(phrase))
      .sort((a, b) => normalize(a.innerText || '').length - normalize(b.innerText || '').length);
    if (!matches.length) return null;
    matches[0].classList.add('ci-box-highlight');
    return matches[0];
  }

  for (const phrase of phrases) {
    let node = markExactPhrase(phrase);
    if (!node) node = markContainingElement(phrase);
    results.push({ phrase, found: Boolean(node) });
  }

  const first = document.querySelector('mark.ci-highlight, .ci-box-highlight');
  if (first) first.scrollIntoView({ block: 'center', inline: 'center', behavior: 'instant' });
  document.documentElement.style.scrollBehavior = 'auto';
  return results;
}
"""


def dismiss_obstructions(page):
    selectors = [
        "button:has-text('Accept')",
        "button:has-text('Accept All')",
        "button:has-text('I Accept')",
        "button:has-text('Agree')",
        "button:has-text('Got it')",
        "button:has-text('Close')",
        "[aria-label='Close']",
        ".onetrust-close-btn-handler",
        "#onetrust-accept-btn-handler",
    ]
    for selector in selectors:
        try:
            locator = page.locator(selector).first
            if locator.count() and locator.is_visible(timeout=700):
                locator.click(timeout=1000)
        except Exception:
            pass


def capture_one(page, source, kind, phrases, output_path):
    page.goto(source["url"], wait_until="domcontentloaded", timeout=65000)
    page.wait_for_timeout(1800)
    dismiss_obstructions(page)
    page.evaluate(
        "() => { document.querySelectorAll('iframe, .ad, .ads, [class*=advert i], [id*=advert i]').forEach(el => el.classList.add('ci-hide')); }"
    )
    result = page.evaluate(HIGHLIGHT_JS, phrases)
    missing = [row["phrase"] for row in result if not row["found"]]
    if missing:
        raise RuntimeError(f"{source['id']} {kind} missing highlights: {missing}")
    page.wait_for_timeout(500)
    page.screenshot(path=str(output_path), full_page=False)


def main():
    DATE_DIR.mkdir(parents=True, exist_ok=True)
    EVIDENCE_DIR.mkdir(parents=True, exist_ok=True)
    chrome = "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"
    with sync_playwright() as p:
        browser = p.chromium.launch(
            executable_path=chrome,
            headless=True,
            args=["--disable-blink-features=AutomationControlled", "--no-sandbox"],
        )
        context = browser.new_context(
            viewport={"width": 1440, "height": 1050},
            device_scale_factor=1.5,
            user_agent=(
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/124.0.0.0 Safari/537.36"
            ),
        )
        page = context.new_page()
        for source in SOURCES:
            capture_one(
                page,
                source,
                "date",
                [source["date"]],
                DATE_DIR / f"source-{source['id']}-date.png",
            )
            capture_one(
                page,
                source,
                "evidence",
                source["evidence"],
                EVIDENCE_DIR / f"source-{source['id']}-evidence-01.png",
            )
            print(f"captured source {source['id']} - {source['name']}")
        browser.close()


if __name__ == "__main__":
    main()

File: competitive_intelligence_reports/head_and_neck_squamous_cell_carcinoma/2026-05-12_1235/scripts/capture_sources.mjs
import fs from "node:fs/promises";
import path from "node:path";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const { chromium } = require("playwright");

const runDir = "/Users/justinyu/Desktop/linkedin-posts/competitive_intelligence_reports/head_and_neck_squamous_cell_carcinoma/2026-05-12_1235";
const dateDir = path.join(runDir, "screenshots/date-verification");
const evidenceDir = path.join(runDir, "screenshots/evidence");

const sources = [
  {
    ref: "01",
    name: "Inhibrx Biosciences",
    url: "https://inhibrxbiosciences.investorroom.com/2026-05-11-Inhibrx-Reports-Interim-Phase-2-Data-for-INBRX-106-in-First-Line-HNSCC-Initial-Results-Demonstrate-Potential-Costimulatory-Benefit-Over-PD-1-Monotherapy",
    date: "May 11, 2026",
    evidence: [
      {
        slug: "orr",
        phrases: [
          "44.0% confirmed Objective Response Rate",
          "44.0% versus 21.4%",
          "53 patients"
        ]
      },
      {
        slug: "next-steps",
        phrases: [
          "progression-free survival data",
          "fourth quarter of 2026",
          "begin the Phase 3 portion"
        ]
      }
    ]
  },
  {
    ref: "02",
    name: "Bicara Therapeutics via GlobeNewswire",
    url: "https://www.globenewswire.com/news-release/2026/05/11/3291783/0/en/bicara-therapeutics-reports-first-quarter-2026-financial-results-and-provides-business-update.html",
    date: "May 11, 2026",
    evidence: [
      {
        slug: "fortifi",
        phrases: [
          "substantially enrolled by the end of the year",
          "interim analysis in mid-2027",
          "loading dose of 1500mg weekly"
        ]
      },
      {
        slug: "asco-cash",
        phrases: [
          "Long-term follow-up data",
          "ASCO 2026",
          "$539.8 million"
        ]
      }
    ]
  },
  {
    ref: "03",
    name: "Corbus Pharmaceuticals",
    url: "https://ir.corbuspharma.com/news-events/press-releases/detail/467/corbus-pharmaceuticals-reports-q1-2026-financial-results-and-provides-a-corporate-update",
    date: "May 12, 2026",
    evidence: [
      {
        slug: "crb701-path",
        phrases: [
          "broad alignment with the FDA",
          "second-line HNSCC this summer",
          "Report monotherapy data"
        ]
      },
      {
        slug: "cash",
        phrases: [
          "$138.2 million",
          "fund operations into 2028"
        ]
      }
    ]
  },
  {
    ref: "04",
    name: "Pyxis Oncology via GlobeNewswire",
    url: "https://www.globenewswire.com/news-release/2026/05/07/3290537/0/en/Pyxis-Oncology-Appoints-Nelson-Azoulay-as-Chief-Business-Officer.html",
    date: "May 07, 2026",
    evidence: [
      {
        slug: "micvo",
        phrases: [
          "advance MICVO",
          "Phase 1 clinical study in patients with recurrent and metastatic head and neck squamous cell carcinoma",
          "Fast Track Designation"
        ]
      }
    ]
  },
  {
    ref: "05",
    name: "PDS Biotechnology",
    url: "https://pdsbiotech.com/index.php/investors/news-center/press-releases/press-releases1/134-2026-news/1037-pds-biotech-announces-conference-call-and-webcast-for-2026-f2026-05-06-050506",
    date: "May 06, 2026",
    evidence: [
      {
        slug: "q1-call-program",
        phrases: [
          "conference call and webcast",
          "Wednesday, May 13, 2026",
          "pivotal clinical trial to advance its lead program in advanced HPV16-positive head and neck squamous cell cancers"
        ]
      }
    ]
  }
];

async function ensureDirs() {
  await fs.mkdir(dateDir, { recursive: true });
  await fs.mkdir(evidenceDir, { recursive: true });
}

async function dismissOverlays(page) {
  const candidates = [
    "Accept All", "Accept all", "Accept", "I Accept", "I agree", "Agree", "Got it",
    "Close", "No thanks", "Reject All", "Continue"
  ];
  for (const label of candidates) {
    try {
      const locator = page.getByText(label, { exact: false }).first();
      if (await locator.isVisible({ timeout: 500 })) {
        await locator.click({ timeout: 1000 });
        await page.waitForTimeout(250);
      }
    } catch {}
  }
}

async function preparePage(page) {
  await page.addStyleTag({
    content: `
      .ci-mark {
        background: rgba(215, 255, 95, 0.78) !important;
        outline: 4px solid #11130f !important;
        box-decoration-break: clone;
        -webkit-box-decoration-break: clone;
        color: inherit !important;
      }
      .ci-box {
        outline: 5px solid #11130f !important;
        box-shadow: 0 0 0 8px rgba(215, 255, 95, 0.45) !important;
        background: rgba(215, 255, 95, 0.22) !important;
      }
      .ci-screenshot-note {
        position: fixed;
        left: 18px;
        top: 18px;
        z-index: 2147483647;
        background: #11130f;
        color: #f6f1e8;
        border: 3px solid #d7ff5f;
        border-radius: 8px;
        padding: 8px 12px;
        font: 700 15px/1.15 Arial, sans-serif;
        max-width: 540px;
      }
    `
  });
}

async function markPhrases(page, phrases, label) {
  const result = await page.evaluate(({ phrases, label }) => {
    document.querySelectorAll(".ci-mark").forEach((node) => {
      const parent = node.parentNode;
      if (!parent) return;
      parent.replaceChild(document.createTextNode(node.textContent || ""), node);
      parent.normalize();
    });
    document.querySelectorAll(".ci-box").forEach((node) => node.classList.remove("ci-box"));
    document.querySelectorAll(".ci-screenshot-note").forEach((node) => node.remove());

    const normalize = (text) => (text || "").replace(/\s+/g, " ").trim();
    const isVisible = (el) => {
      if (!el || !(el instanceof Element)) return false;
      const style = window.getComputedStyle(el);
      const rect = el.getBoundingClientRect();
      return style.visibility !== "hidden" && style.display !== "none" && rect.width > 0 && rect.height > 0;
    };

    const marked = [];
    const walker = document.createTreeWalker(document.body, NodeFilter.SHOW_TEXT, {
      acceptNode(node) {
        if (!normalize(node.nodeValue)) return NodeFilter.FILTER_REJECT;
        const parent = node.parentElement;
        if (!isVisible(parent)) return NodeFilter.FILTER_REJECT;
        if (["SCRIPT", "STYLE", "NOSCRIPT"].includes(parent.tagName)) return NodeFilter.FILTER_REJECT;
        return NodeFilter.FILTER_ACCEPT;
      }
    });

    const textNodes = [];
    while (walker.nextNode()) textNodes.push(walker.currentNode);

    for (const phrase of phrases) {
      const target = normalize(phrase);
      let found = false;

      for (const node of textNodes) {
        if (!node.parentNode) continue;
        const normalizedNode = normalize(node.nodeValue);
        if (!normalizedNode.includes(target)) continue;

        const raw = node.nodeValue || "";
        const idx = raw.replace(/\u00a0/g, " ").toLowerCase().indexOf(phrase.toLowerCase());
        if (idx >= 0) {
          const before = document.createTextNode(raw.slice(0, idx));
          const span = document.createElement("span");
          span.className = "ci-mark";
          span.textContent = raw.slice(idx, idx + phrase.length);
          const after = document.createTextNode(raw.slice(idx + phrase.length));
          node.parentNode.insertBefore(before, node);
          node.parentNode.insertBefore(span, node);
          node.parentNode.insertBefore(after, node);
          node.parentNode.removeChild(node);
          marked.push({ phrase, mode: "span" });
          found = true;
          break;
        }
      }

      if (!found) {
        const candidates = [...document.querySelectorAll("p, li, h1, h2, h3, h4, div, span, td")].filter((el) =>
          isVisible(el) && normalize(el.innerText).includes(target)
        );
        candidates.sort((a, b) => normalize(a.innerText).length - normalize(b.innerText).length);
        if (candidates[0]) {
          candidates[0].classList.add("ci-box");
          marked.push({ phrase, mode: "box" });
          found = true;
        }
      }

      if (!found) marked.push({ phrase, mode: "missing" });
    }

    const note = document.createElement("div");
    note.className = "ci-screenshot-note";
    note.textContent = label;
    document.body.appendChild(note);

    const first = document.querySelector(".ci-mark, .ci-box");
    if (first) first.scrollIntoView({ block: "center", inline: "nearest" });
    return marked;
  }, { phrases, label });

  await page.waitForTimeout(650);
  return result;
}

async function captureItem(page, source) {
  await page.goto(source.url, { waitUntil: "domcontentloaded", timeout: 90000 });
  await page.waitForTimeout(1800);
  await dismissOverlays(page);
  await preparePage(page);

  const dateResult = await markPhrases(page, [source.date], `Reference ${Number(source.ref)} - date verification`);
  await page.screenshot({
    path: path.join(dateDir, `source-${source.ref}-date.png`),
    fullPage: false
  });

  const evidenceResults = [];
  for (let i = 0; i < source.evidence.length; i += 1) {
    const item = source.evidence[i];
    const result = await markPhrases(page, item.phrases, `Reference ${Number(source.ref)} - evidence ${i + 1}`);
    await page.screenshot({
      path: path.join(evidenceDir, `source-${source.ref}-evidence-${String(i + 1).padStart(2, "0")}-${item.slug}.png`),
      fullPage: false
    });
    evidenceResults.push({ item, result });
  }

  return { ref: source.ref, name: source.name, dateResult, evidenceResults };
}

await ensureDirs();
const browser = await chromium.launch({ headless: true, args: ["--no-sandbox", "--disable-dev-shm-usage"] });
const context = await browser.newContext({
  viewport: { width: 1440, height: 1040 },
  deviceScaleFactor: 1,
  userAgent: "codex-ci/1.0 contact justinyu@example.com"
});
const page = await context.newPage();
const log = [];
for (const source of sources) {
  try {
    log.push(await captureItem(page, source));
  } catch (error) {
    log.push({ ref: source.ref, name: source.name, error: String(error?.stack || error) });
  }
}
await browser.close();
console.log(JSON.stringify(log, null, 2));
